src/platform_impl/macos/window_delegate.rs

```diff
@@ -275,10 +288,18 @@ extern "C" fn window_did_move(this: &Object, _: Sel, _: id) {
 extern "C" fn window_did_change_screen(this: &Object, _: Sel, _: id) {
     trace!("Triggered `windowDidChangeScreen:`");
     with_state(this, |state| {
-        let dpi_factor = unsafe { NSWindow::backingScaleFactor(*state.ns_window) } as f64;
-        if state.previous_dpi_factor != dpi_factor {
-            state.previous_dpi_factor = dpi_factor;
-            state.emit_event(WindowEvent::HiDpiFactorChanged(dpi_factor));
+        let hidpi_factor = unsafe { NSWindow::backingScaleFactor(*state.ns_window) } as f64;
+        let ns_size = unsafe { NSWindow::contentSize(*state.ns_window) };
+        if state.previous_dpi_factor != hidpi_factor {
+            state.previous_dpi_factor = hidpi_factor;
+            state.emit_event(
+                WindowEvent::HiDpiFactorChanged {
+                    hidpi_factor,
+                    new_inner_size: &mut Some(inner_size(ns_size, hidpi_factor)),
+                }
+                .to_static()
+                .unwrap(),
+            );
```

Comment:
WRT `HiDpiFactorChanged` (these comments also apply to the other places in the code where it gets dispatched):

The `new_inner_size` field is supposed to take a mutable reference to the new window size suggested by the OS. That size can then be modified by the user if they would like to resize the window in a different way in response to the DPI change, and the window's size after the DPI change gets set to whatever value is pointed to by `new_inner_size` after the event handler. If the OS doesn't suggest a size, we set it to `None`, and if the user wants to ignore any sizing changes, they can set `new_inner_size` to `None` in their callback and we should leave the window's size as it was before the DPI factor change.

With that in mind, there are a couple concrete problems I can point to in the current implementation that must be solved before this gets merged:

1. `new_inner_size` must reference a value that later gets read by the code dispatching the event. Right now, the event handler doesn't do that, so any new value the user passes to `new_inner_size` is getting ignored.
2. `HiDpiFactorChanged` cannot use `emit_event`. This is because `emit_event` queues the event for dispatch, rather than dispatching it immediately, and thus requires an event with a `'static` lifetime. Skimming the code it doesn't look like the function you need exists right now, but you should be able to adapt the code in [`emit_resize_event`](https://github.com/rust-windowing/winit/blob/3a449af496485e0e106b5b01234480507121aa94/src/platform_impl/macos/window_delegate.rs#L84-L92), since that code dispatches the event without going through an event queue.

As far as testing this goes, it'll probably be easiest if you've got another monitor somewhere in your house that you can hook up, since it seems like that's the main place this event with get dispatched from. If that's not an option, it's worth looking into where [`windowDidChangeBackingProperties`](https://developer.apple.com/documentation/appkit/nswindowdelegate/1419517-windowdidchangebackingproperties) gets dispatched since that's the other place this event can get emitted from, but how exactly you trigger that seems less clear from a cursory glance.

Also, it may be helpful to look at the [win32 implementation for reference](https://github.com/rust-windowing/winit/blob/3a449af496485e0e106b5b01234480507121aa94/src/platform_impl/windows/event_loop.rs#L1568-L1590) if you'd like a concrete code example.

As always, if any part of that was unclear or if you've got any questions, feel free to ask and I'll answer them to the best of my ability.

Reply:
Hi! Thanks for adding some clarity. I was quite confused why `HiDpiFactorChanged` should've been `'static`.
Do we get a new `WindowDelegateState` every time dpi factor changes?

Reply:
@vbogaevsky Doesn't look like it. I think `WindowDelegateState` is created alongside the window, and the `HiDpiFactorChanged` event emitted there is just used to notify the event loop that the window starts out with a non-`1.0` DPI factor.

Reply:
@Osspial `emit_resize_event` utilises `AppState::send_event_immediately` which as `emit_event` expects event to be `'static` too. As I understand, it works for  `emit_resize_event` because `WindowEvent::Resized` holds no references, so it can be evaluated as `'static`, however we have to give `WindowEvent::HiDpiFactorChanged` a reference to an `Option` holding `new_inner_size`.
The only logical solution I see is to make `WindowDelegateState` `'static', then it will be holding a '`static' value which we will be able to reference.

Reply:
I hope I found a solution to this problem with `'static`